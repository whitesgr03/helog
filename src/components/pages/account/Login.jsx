// Packages
import { Navigate, useOutletContext } from "react-router-dom";
import { useState } from "react";

// style
import image from "../../../styles/utils/image.module.css";
import style from "../../../styles/pages/account/login.module.css";

// Components
import { Loading } from "../../layout/Loading";

// Assets
import googleIcon from "../../../assets/google.png";
import facebookIcon from "../../../assets/facebook.png";

export const Login = () => {
	const { user } = useOutletContext();

	const [loading, setLoading] = useState(false);

	const handleSocialLogin = async provider => {
		setLoading(true);

		window.location.assign(
			`${import.meta.env.VITE_RESOURCE_URL}/account/login/${provider}`
		);

		setLoading(false);
	};

	return (
		<>
			{user ? (
				<Navigate to="/" replace={true} />
			) : (
				<div className={style.account}>
					<h3 className={style.title}>User Sign in</h3>
					<div className={style.container}>
						{loading && <Loading text={"Submitting..."} />}
						<div className={style.federation}>
							<button
								className={style["federation-button"]}
								onClick={() => handleSocialLogin("google")}
							>
								<div className={`${image} ${style.google}`}>
									<img src={googleIcon} alt="Google icon" />
								</div>
								Sign in with Google
							</button>
							<button
								className={style["federation-button"]}
								onClick={() => handleSocialLogin("facebook")}
							>
								<div className={`${image} ${style.facebook}`}>
									<img
										src={facebookIcon}
										alt="Facebook icon"
									/>
								</div>
								Sign in with Facebook
							</button>
						</div>
					</div>
				</div>
			)}
		</>
	);
};
