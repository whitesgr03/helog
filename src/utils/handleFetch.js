const handleFetch = async (url, option) => {
	const response = await fetch(url, { ...option });

	const result = await response.json();

	!result.success &&
		(() => {
			throw new Error("Fetch Error", {
				cause: result.errors ?? result.message,
			});
		})();

	return result.data;
};

export { handleFetch };
