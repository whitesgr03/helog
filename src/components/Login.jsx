import { useState } from "react";
import { useOutletContext, Link, useNavigate } from "react-router-dom";

import style from "../styles/Login.module.css";
import form from "../styles/utils/form.module.css";
import button from "../styles/utils/button.module.css";

import handleFetch from "../utils/handleFetch";

const POST_LOGIN_URL = "http://localhost:3000/blog/users/login";

const Login = () => {
	const { setToken } = useOutletContext();
	const [errors, setErrors] = useState({});
	const navigate = useNavigate();

	const handelSubmit = async e => {
		e.preventDefault();
		const formData = new FormData(e.target);
		const formProps = Object.fromEntries(formData);

		const fetchOption = {
			method: "POST",
			headers: {
				"Content-Type": "application/json",
			},
		};
		try {
			const data = await handleFetch(POST_LOGIN_URL, {
				...fetchOption,
				body: JSON.stringify(formProps),
			});

			localStorage.setItem("token", JSON.stringify(data));
			setToken(data.token);
			navigate("/", { replace: true });
		} catch (err) {
			const obj = {};
			for (const error of err.cause) {
				obj[error.field] = error.message;
			}
			setErrors(obj);
		}
	};

	return (
		<div className={style.login}>
			<h3 className={style.title}>Sign In</h3>
			<div className={style.formWrap}>
				<form className={form.content} onSubmit={handelSubmit}>
					<div>
						<label
							htmlFor="loginEmail"
							className={`${errors.email ? form.error : ""}`}
						>
							Email
							<input type="email" id="loginEmail" name="email" />
						</label>
						<span>
							{errors.email
								? errors.email
								: "This is a placeholder"}
						</span>
					</div>
					<div>
						<label
							htmlFor="loginPassword"
							className={`${errors.password ? form.error : ""}`}
						>
							Password
							<input
								type="password"
								id="loginPassword"
								name="password"
							/>
						</label>
						<span>
							{errors.password
								? errors.password
								: "This is a placeholder"}
						</span>
					</div>
					<button type="submit" className={button.success}>
						Login
					</button>
				</form>
				<div className={style.linkWrap}>
					<p>New to HeLog?</p>
					<Link to="/users/register">Create an account</Link>
				</div>
			</div>
		</div>
	);
};

export default Login;
