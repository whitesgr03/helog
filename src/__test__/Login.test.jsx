import { vi } from "vitest";
import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { BrowserRouter, Outlet, Route, Routes } from "react-router-dom";

import Login from "../components/Login";

import handleFetch from "../utils/handleFetch";

vi.mock("../utils/handleFetch");

/* eslint-disable react/prop-types */
const Providers = ({ children, ...props }) => {
	return (
		<BrowserRouter>
			<Routes>
				<Route element={<Outlet context={{ ...props }} />}>
					<Route index element={children} />
				</Route>
			</Routes>
		</BrowserRouter>
	);
};

describe("Login component", () => {
	it("should render error messages if the validation field fails when submitting.", async () => {
		const user = userEvent.setup();
		const mockOutContext = {
			setToken: vi.fn(),
		};

		render(
			<Providers {...mockOutContext}>
				<Login />
			</Providers>
		);

		const submitBtn = screen.getByRole("button", { name: "Login" });

		await user.click(submitBtn);

		const defaultErrorMessages = screen.queryAllByText(
			"Error message placeholder"
		);

		expect(defaultErrorMessages.length).toEqual(0);
	});
	it("should login successfully if the validation field is successful when submitting.", async () => {
		const user = userEvent.setup();
		const mockOutletContext = {
			setToken: vi.fn(),
		};

		handleFetch.mockReturnValueOnce({ data: [] });

		render(
			<Providers {...mockOutletContext}>
				<Login />
			</Providers>
		);

		const emailInput = screen.getByLabelText("Email");
		const passwordInput = screen.getByLabelText("Password");
		const submitBtn = screen.getByRole("button", { name: "Login" });

		await user.type(emailInput, "example@gmail.com");
		await user.type(passwordInput, "1!Qwerty");

		await user.click(submitBtn);

		expect(handleFetch).toBeCalledTimes(1);
		expect(mockOutletContext.setToken).toBeCalledTimes(1);
	});
	it("should continue validating fields when entering again after a failed submission validation.", async () => {
		const user = userEvent.setup();
		const mockOutContext = {
			setToken: vi.fn(),
		};

		render(
			<Providers {...mockOutContext}>
				<Login />
			</Providers>
		);

		const submitBtn = screen.getByRole("button", { name: "Login" });
		const emailInput = screen.getByLabelText("Email");
		const passwordInput = screen.getByLabelText("Password");

		await user.click(submitBtn);

		let defaultErrorMessages = screen.queryAllByText(
			"Error message placeholder"
		);

		expect(defaultErrorMessages.length).toEqual(0);

		await user.type(emailInput, "example@gmail.com");
		await user.type(passwordInput, "1!Qwerty");

		defaultErrorMessages = await screen.findAllByText(
			"Error message placeholder"
		);

		expect(defaultErrorMessages.length).toEqual(2);
	});
	it("should render error messages if handle fetch fails", async () => {
		const user = userEvent.setup();
		const mockOutletContext = {
			setToken: vi.fn(),
		};

		handleFetch.mockImplementationOnce(() => {
			throw new Error("Fetch Error", {
				cause: [
					{
						field: "email",
						message: "error",
					},
					{
						field: "password",
						message: "error",
					},
				],
			});
		});

		render(
			<Providers {...mockOutletContext}>
				<Login />
			</Providers>
		);

		const emailInput = screen.getByLabelText("Email");
		const passwordInput = screen.getByLabelText("Password");
		const submitBtn = screen.getByRole("button", { name: "Login" });

		await user.type(emailInput, "example@gmail.com");
		await user.type(passwordInput, "1!Qwerty");

		await user.click(submitBtn);

		const defaultErrorMessages = screen.queryAllByText(
			"Error message placeholder"
		);

		expect(defaultErrorMessages.length).toEqual(0);
		expect(handleFetch).toBeCalledTimes(1);
	});
});
