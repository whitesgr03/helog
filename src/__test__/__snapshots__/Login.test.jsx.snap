// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login component > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="_login_b06c4f"
  >
    <h3
      class="title"
    >
      Sign In
    </h3>
    <div
      class="formWrap"
    >
      <form>
        <label
          for="email"
        >
          Email
        </label>
        <input
          id="email"
          name="email"
          type="email"
        />
        <label
          for="password"
        >
          Password
        </label>
        <input
          id="password"
          name="password"
          type="password"
        />
        <button
          class="submitBtn"
          type="submit"
        >
          Login
        </button>
      </form>
      <div
        class="_linkWrap_b06c4f"
      >
        <p>
          New to HeLog?
        </p>
        <a
          class="_link_b06c4f"
          href="/user/register"
        >
          Create an account
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
